### PR TITLE
Fix completion in func exp and class exp

### PR DESCRIPTION
**Pull request by yuit**
Fix issue #3231 and add support for class expression

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -2808,6 +2811,22 @@ namespace ts {
...
                 }
 
+                // Special case for function expression and class expression because despite sometimes having a name, the binder
+                // binds them to a symbol with the name "__function" and "__class" respectively. However, for completion entry, we want
+                // to display its declared name rather than "__function" and "__class".
+                //      var x = function foo () {
+                //          fo$  <- completion list should contain local name "foo"
+                //      }
+                //      foo$ <- completion list should not contain "foo"
+                if (displayName === "__function" || displayName === "__class") {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: We have a `getDeclaredName` which I think could be used in lieu of this logic.

File: src/services/services.ts [hunk 2]
```diff
@@ -3531,7 +3550,8 @@ namespace ts {
         function getSymbolKind(symbol: Symbol, location: Node): string {
             let flags = symbol.getFlags();
 
-            if (flags & SymbolFlags.Class) return ScriptElementKind.classElement;
+            if (flags & SymbolFlags.Class) return ts.forEach(symbol.declarations, d => d.kind === SyntaxKind.ClassExpression) ?
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Maybe just use `firstOrUndefined` and see if that's a class expression; those can't have multiple declarations

Comment (DanielRosenwasser) [inline src/services/services.ts]: Also, any logic simpler than the other cases should have its own block body

File: src/services/services.ts [hunk 3]
```diff
@@ -1424,6 +1424,9 @@ namespace ts {
         // class X {}
         export const classElement = "class";
 
+        // var y = class X {}
+        export const localClassElement = "local class";
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: I feel like "class" would suffice but I can get behind this

File: src/services/services.ts [hunk 4]
```diff
@@ -3740,7 +3760,16 @@ namespace ts {
                 }
             }
             if (symbolFlags & SymbolFlags.Class && !hasAddedSymbolInfo) {
-                displayParts.push(keywordPart(SyntaxKind.ClassKeyword));
+                // Special case for class expression because we would like the entry detail to indicate that
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: expression -> expressions

Comment (DanielRosenwasser) [inline src/services/services.ts]: Is this specific to entry details? I thought quick info used it to. I would just say "we would like to indicate that"

**Comment by DanielRosenwasser** [comment]
@yuit why was this closed?

**Comment by yuit** [comment]
@DanielRosenwasser  my PR is snapped using master from like 2 weeks ago :(.

File: src/services/services.ts (hunk 1 above)
Comment (yuit) [inline src/services/services.ts]: ```getDeclaredName``` doesn't do the right thing for functionExpression. Talk with @mhegazy, he says there will be some rewrite on that. So I don't want to have all the change in this PR

File: src/services/services.ts (hunk 3 above)
Comment (yuit) [inline src/services/services.ts]: I want to keep it consistent with function expression

File: src/services/services.ts (hunk 4 above)
Comment (yuit) [inline src/services/services.ts]: you are right. It should in quick info. I will change that

**Comment by DanielRosenwasser** [comment]
You can still merge in from `master`, and have things reviewed in the files changed list, unless it's really just that bad.

**Comment by yuit** [comment]
It is huge . Anyhow, I cherry-picked some commit and incorporate your comments. Sorry for confusion here

**Pull request closed without merging**